kuma/javascript/src/header/main-menu.jsx

```diff
@@ -0,0 +1,266 @@
+     */
+    function desktopInteractionHandlers(menu) {
+        menu.addEventListener('mouseover', (event: MouseEvent) => {
+            let currentTarget = event.target;
+            /* currentTarget instanceof 'HTMLButtonElement' is added
+               to keep Flow happy: https://github.com/facebook/flow/issues/218#issuecomment-74119319 */
+            if (
+                currentTarget instanceof HTMLButtonElement &&
+                currentTarget.classList.contains('top-level-entry')
+            ) {
+                if (currentTarget.nextElementSibling) {
+                    currentTarget.nextElementSibling.setAttribute(
+                        'aria-hidden',
+                        'false'
+                    );
+                }
+            }
+        });
+
+        menu.addEventListener('mouseout', (event: MouseEvent) => {
+            let currentTarget = event.target;
+            if (
+                currentTarget instanceof HTMLButtonElement &&
+                currentTarget.classList.contains('top-level-entry')
+            ) {
+                if (currentTarget.nextElementSibling) {
+                    currentTarget.nextElementSibling.setAttribute(
+                        'aria-hidden',
+                        'false'
+                    );
+                }
+            }
+        });
+    }
+
+    /**
+     * Handles touch events on mobile devices
+     * @param {Object} menu - The current HTML nav element
+     */
+    function mobileInteractionHandlers(menu) {
+        menu.addEventListener('touchend', (event: TouchEvent) => {
+            event.stopImmediatePropagation();
+            let currentTarget = event.target;
+            if (
+                currentTarget instanceof HTMLButtonElement &&
+                currentTarget.classList.contains('top-level-entry')
+            ) {
+                let subMenu = currentTarget.nextElementSibling;
+
+                if (subMenu) {
+                    let currentAriaState = subMenu.getAttribute('aria-hidden');
+                    let newAriaState =
+                        currentAriaState === 'true' ? 'false' : 'true';
+                    subMenu.setAttribute('aria-hidden', newAriaState);
+                    subMenu.classList.toggle('show');
+                }
+            }
+        });
+    }
+
+    /**
+     * Hide the visible submenu in main navigation
+     */
+    function hideSubMenus() {
+        let visibleSubMenu = document.querySelector(
+            'li ul[aria-hidden="false"]'
+        );
+        if (visibleSubMenu) {
+            visibleSubMenu.setAttribute('aria-hidden', 'true');
+            visibleSubMenu.classList.toggle('show');
+        }
+    }
+
+    useEffect(() => {
+        let menu = document.getElementById('header-main-nav');
+        var mediaQuery = window.matchMedia('(min-width: 47.9375em)');
+        if (menu) {
+            if (mediaQuery.matches) {
+                desktopInteractionHandlers(menu);
```

Comment:
I'm pretty sure that this code is going to end up registering the same event handlers multiple times. Each time the menu is rendered, this effect function is going to run, but that doesn't mean that there will be a new menu element each time, so you're likely to end up with a memory leak here (and also below for the touch event handlers).

One approach would be to define the required event handlers directly in the rendering code, rather than adding them on with this effect.

The other approach would be to write corresponding functions for removing the event handlers.  If useEffect() returns a function  then that function will be used as the cleanup function. If you do that then the cleanup function will always be called before the effect function is called again.